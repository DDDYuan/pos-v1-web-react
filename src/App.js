import React, { PureComponent } from "react";
import { getAllItems, getAllPromotions } from "./database";

class App extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      items: getAllItems(),
      promotions: getAllPromotions(),
      listPage: true,
      cart: []
    };
  }

  goToCartView = () => {
    this.setState({ listPage: false });
  };

  goToListView = () => {
    this.setState({ listPage: true });
  };

  addToCart = item => {
    this.setState({ cart: [...this.state.cart, item] });
  };

  logCart = () => {
    const result = [];
    this.state.cart.forEach(item => {
      result.push(`${item.barcode}-${item.amount}`);
    });
    console.log(result);
  };

  render() {
    if (this.state.listPage) {
      return (
        <ItemList
          items={this.state.items}
          promotions={this.state.promotions}
          changePage={this.goToCartView}
          addToCart={this.addToCart}
        />
      );
    }
    return (
      <div>
        <ShoppingCart
          items={this.state.cart}
          changePage={this.goToListView}
          logCart={this.logCart}
        />
      </div>
    );
  }
}

class ItemList extends PureComponent {
  getItemByCode = code => {
    return this.props.items.find(item => item.barcode === code);
  };

  onAdd = event => {
    const barcode = event.currentTarget.parentElement.parentElement.getAttribute(
      "barcode"
    );
    const amount = parseInt(
      event.currentTarget.parentElement.parentElement.firstChild.value,
      10
    );
    if (amount > 0) {
      this.props.addToCart({
        ...this.getItemByCode(barcode),
        amount: amount
      });
      event.currentTarget.parentElement.parentElement.firstChild.value = 0;
      this.setOkStyle(event.currentTarget);
    }
  };

  setOkStyle = button => {
    const originalStyle = button.className;
    const originalText = button.textContent;
    button.className = "btn btn-success btn-sm";
    button.textContent = "OK !";

    setTimeout(() => {
      button.className = originalStyle;
      button.textContent = originalText;
    }, 1000);
  };

  render() {
    return (
      <div>
        <button
          className="btn btn-lg btn-block btn-outline-info"
          onClick={this.props.changePage}
        >
          GO TO SHOPPING CART
        </button>
        <h3 className="bg-light mb-0">ITEMS LIST</h3>
        <ul className="list-group">
          {this.props.items.map(item => (
            <li key={item.barcode}>
              <div className="list-group-item d-flex justify-content-between align-items-center row">
                <div className="col">{item.name}</div>
                <div className="badge badge-info col-1">
                  {item.price}元/{item.unit}
                </div>
                <div className="col-3 input-group" barcode={item.barcode}>
                  <input
                    type="number"
                    min="0"
                    defaultValue="0"
                    className="form-control"
                  />
                  <div className="input-group-append">
                    <button
                      onClick={this.onAdd}
                      className="btn btn-primary btn-sm"
                    >
                      ADD
                    </button>
                  </div>
                </div>
              </div>
            </li>
          ))}
        </ul>
        <hr />
        <h3 className="bg-light mb-0">PROMOTIONS LIST</h3>
        <ul className="list-group">
          {this.props.promotions.map(item =>
            item.barcodes.map(barcode => (
              <li key={item.type + barcode}>
                <div className="list-group-item d-flex justify-content-between align-items-center row">
                  <div className="col">{this.getItemByCode(barcode).name}</div>
                  <div className="badge badge-success col-3">{item.type}</div>
                </div>
              </li>
            ))
          )}
        </ul>
      </div>
    );
  }
}

class ShoppingCart extends PureComponent {
  getTotalAmount = () => {
    let amount = 0;
    this.props.items.forEach(item => {
      amount += item.price * item.amount;
    });
    return amount;
  };

  render() {
    return (
      <div>
        <button
          className="btn btn-lg btn-block btn-outline-info"
          onClick={this.props.changePage}
        >
          GO TO ITEM LIST
        </button>
        <h3>SHOPPING CART</h3>
        <ul className="list-group">
          {this.props.items.map(item => (
            <li key={item.barcode}>
              <div className="list-group-item d-flex justify-content-between align-items-center row">
                <div className="col-4">{item.name}</div>
                <div className="badge badge-info col-1">
                  {item.price}元/{item.unit}
                </div>
                <div className="col-4 text-right">
                  共 {item.amount} {item.unit}
                </div>
                <div className="col text-right">
                  小计： {item.amount * item.price} 元
                </div>
              </div>
            </li>
          ))}
          <li key="total">
            <div className="list-group-item d-flex justify-content-between align-items-center row">
              <div className="col">总计：</div>
              <div className="col-2 text-right">{this.getTotalAmount()}元</div>
            </div>
          </li>
        </ul>
        <button
          className="btn btn-lg btn-block btn-success"
          onClick={this.props.logCart}
        >
          CONFIRM
        </button>
      </div>
    );
  }
}

export default App;
