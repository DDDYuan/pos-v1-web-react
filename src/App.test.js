import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import Renderer from "react-test-renderer";
import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

describe("Application", () => {
  beforeEach(() => {
    Enzyme.configure({ adapter: new Adapter() });
  });
  it("should render without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("should render App correctly", () => {
    const tree = Renderer.create(<App />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("should add item to cart correctly", () => {
    const app = mount(<App />);
    const itemSelector = "div[barcode*='001']";

    let log;
    console.log = jest.fn(content => (log = content));

    expect(
      app
        .find(itemSelector)
        .childAt(1)
        .text()
    ).toEqual("ADD");

    app.find(`${itemSelector} input`).instance().value = 1;
    app.find(`${itemSelector} button`).simulate("click");
    app.find(".btn-block").simulate("click");
    expect(app.find("div.col-2").text()).toEqual("3å…ƒ");

    app.find(".btn-success").simulate("click");
    expect(log).toEqual(["ITEM000001-1"]);
  });
});
