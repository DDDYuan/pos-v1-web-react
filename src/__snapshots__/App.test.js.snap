// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application should render App correctly 1`] = `
<div
  className="row justify-content-center"
>
  <div
    className="col-6"
  >
    <button
      className="btn btn-lg btn-block btn-outline-info"
      onClick={[Function]}
    >
      GO TO SHOPPING CART
    </button>
    <h3
      className="mb-0"
    >
      ITEMS LIST
    </h3>
    <ul
      className="list-group"
    >
      <li>
        <div
          className="list-group-item d-flex justify-content-between align-items-center"
        >
          <div
            className="col"
          >
            可口可乐
          </div>
          <div
            className="badge badge-info col-1"
          >
            3
            元/
            瓶
          </div>
          <div
            barcode="ITEM000000"
            className="col-3 input-group"
          >
            <input
              className="form-control"
              defaultValue="0"
              min="0"
              type="number"
            />
            <div
              className="input-group-append"
            >
              <button
                className="btn btn-primary btn-sm"
                onClick={[Function]}
              >
                ADD
              </button>
            </div>
          </div>
        </div>
      </li>
      <li>
        <div
          className="list-group-item d-flex justify-content-between align-items-center"
        >
          <div
            className="col"
          >
            雪碧
          </div>
          <div
            className="badge badge-info col-1"
          >
            3
            元/
            瓶
          </div>
          <div
            barcode="ITEM000001"
            className="col-3 input-group"
          >
            <input
              className="form-control"
              defaultValue="0"
              min="0"
              type="number"
            />
            <div
              className="input-group-append"
            >
              <button
                className="btn btn-primary btn-sm"
                onClick={[Function]}
              >
                ADD
              </button>
            </div>
          </div>
        </div>
      </li>
      <li>
        <div
          className="list-group-item d-flex justify-content-between align-items-center"
        >
          <div
            className="col"
          >
            苹果
          </div>
          <div
            className="badge badge-info col-1"
          >
            5.5
            元/
            斤
          </div>
          <div
            barcode="ITEM000002"
            className="col-3 input-group"
          >
            <input
              className="form-control"
              defaultValue="0"
              min="0"
              type="number"
            />
            <div
              className="input-group-append"
            >
              <button
                className="btn btn-primary btn-sm"
                onClick={[Function]}
              >
                ADD
              </button>
            </div>
          </div>
        </div>
      </li>
      <li>
        <div
          className="list-group-item d-flex justify-content-between align-items-center"
        >
          <div
            className="col"
          >
            荔枝
          </div>
          <div
            className="badge badge-info col-1"
          >
            15
            元/
            斤
          </div>
          <div
            barcode="ITEM000003"
            className="col-3 input-group"
          >
            <input
              className="form-control"
              defaultValue="0"
              min="0"
              type="number"
            />
            <div
              className="input-group-append"
            >
              <button
                className="btn btn-primary btn-sm"
                onClick={[Function]}
              >
                ADD
              </button>
            </div>
          </div>
        </div>
      </li>
      <li>
        <div
          className="list-group-item d-flex justify-content-between align-items-center"
        >
          <div
            className="col"
          >
            电池
          </div>
          <div
            className="badge badge-info col-1"
          >
            2
            元/
            个
          </div>
          <div
            barcode="ITEM000004"
            className="col-3 input-group"
          >
            <input
              className="form-control"
              defaultValue="0"
              min="0"
              type="number"
            />
            <div
              className="input-group-append"
            >
              <button
                className="btn btn-primary btn-sm"
                onClick={[Function]}
              >
                ADD
              </button>
            </div>
          </div>
        </div>
      </li>
      <li>
        <div
          className="list-group-item d-flex justify-content-between align-items-center"
        >
          <div
            className="col"
          >
            方便面
          </div>
          <div
            className="badge badge-info col-1"
          >
            4.5
            元/
            袋
          </div>
          <div
            barcode="ITEM000005"
            className="col-3 input-group"
          >
            <input
              className="form-control"
              defaultValue="0"
              min="0"
              type="number"
            />
            <div
              className="input-group-append"
            >
              <button
                className="btn btn-primary btn-sm"
                onClick={[Function]}
              >
                ADD
              </button>
            </div>
          </div>
        </div>
      </li>
    </ul>
    <hr />
    <h3
      className="mb-0"
    >
      PROMOTIONS LIST
    </h3>
    <ul
      className="list-group"
    >
      <li>
        <div
          className="list-group-item d-flex justify-content-between align-items-center"
        >
          <div
            className="col"
          >
            可口可乐
          </div>
          <div
            className="badge badge-success col-3"
          >
            BUY_TWO_GET_ONE_FREE
          </div>
        </div>
      </li>
      <li>
        <div
          className="list-group-item d-flex justify-content-between align-items-center"
        >
          <div
            className="col"
          >
            雪碧
          </div>
          <div
            className="badge badge-success col-3"
          >
            BUY_TWO_GET_ONE_FREE
          </div>
        </div>
      </li>
      <li>
        <div
          className="list-group-item d-flex justify-content-between align-items-center"
        >
          <div
            className="col"
          >
            方便面
          </div>
          <div
            className="badge badge-success col-3"
          >
            BUY_TWO_GET_ONE_FREE
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
